File: Searching/Binary_search/binary_search.cpp
#include <iostream>
#include <vector>


void binary_search(const std::vector<int>& arr, const int& k)
{

	int low = 0;
	int high = arr.size() - 1;
	bool found = false;
	int index_of_key = -1;

	if ((high == low) && (arr[0] == k))
	{
		found = true;
		index_of_key = 0;
	}

	int mid = (high + low) / 2;

	while ((low <= high) && (!found))
	{

		if (arr[mid] == k)
		{

			found = true;
			index_of_key = mid;

		}
		else
		{ 

			if (k < arr[mid])
				high = mid - 1;
			else 
				low = mid + 1;

			mid = (high + low) / 2;
		}
	}
	std::cout << "Element " << k;

	if (found)
		std::cout << " found at index: " << index_of_key << "." << std::endl;
	else
		std::cout << " not found." << std::endl;

}


int main()
{

	std::vector<int> darray;

	int vertices_length;

	int element;

	int key;
	std::cout << "Enter the no of elements: " << std::endl;

	std::cin >> vertices_length;

	std::cout << "Enter the element(s): " << std::endl;
	while (vertices_length--)
	{

		std::cin >> element;
		darray.push_back(element);

	}
	std::cout << "Enter the element to be searched: " << std::endl;
	std::cin >> key;

	binary_search(darray, key);


	return 0;
}

File: Searching/binary_search.py
arr = []


def accept_user_input():
    
    global arr
    
    print("Enter the elements: ", end="\n")
    arr = list(input().split())
    arr = [int(i) for i in arr]
    

def binary_search(arr, key):
    
    low = 0
    high = len(arr) - 1
    found = False
    index_of_element = -1
    
    if ((high == low) and (arr[0] == key)):
        
        index_of_element = 0
        found = True
    
    mid = (high+low) // 2
   
    while ((low <= high) and (not found)):
    
        if arr[mid] == key:
            
            found = True
            index_of_element = mid
            
        else:
            
            if key < arr[mid]:
                high = mid - 1
            elif key > arr[mid]:
                low = mid + 1
            
            mid = (low+high) // 2    
                       
    print("Element", key, end="")
    
    if found:
        
        print(" found at index: ", index_of_element, end="\n")
    
    else:
        
        print(" not found.", end="\n")                    
        

accept_user_input()

key = int(input("Enter the search element: "))
    
binary_search(arr, key)    
